webserver.py

```diff
@@ -176,216 +308,76 @@ def game_config():
-    game_state['resources']['fertilizer'] += seed_data['yield']['fertilizerYield']
-    if seed_count > GAME_CONFIG['max_seed_count']:
-        overflow = seed_count - GAME_CONFIG['max_seed_count']
-        seed_count = GAME_CONFIG['max_seed_count']
-        game_state['resources']['cash'] += seed_data['yield']['cashYield'] * overflow
-    game_state['seedCounts'][seed_type] = seed_count
-    plot['seedType'] = 0
-
-    recipe_list = []
-    for recipe_id in RECIPE_CONFIG['recipes']:
-        recipe_list.append(recipe_id)
-    for resource in GAME_CONFIG['starting_resources']:
-        if game_state['resources'][resource] == 0:
-            for recipe_id in RECIPE_CONFIG['recipes']:
-                if RECIPE_CONFIG['recipes'][recipe_id][resource + 'Cost'] > 0:
-                    recipe_list.remove(recipe_id)
-    for recipe_id in recipe_list:
-        if recipe_id not in game_state['recipes']:
-            game_state['recipes'].append(recipe_id)
-
-    save_state(data['slug'], game_state)
-
-    message = 'Harvested a %s.' % GAME_CONFIG['seeds'][seed_type]['name']
-    if bonus == 1:
-        message += ' Got 1 bonus seed!'
-    if bonus > 1:
-        message += ' Got %s bonus seeds!' % bonus
-
+    game_state = GameState.load(data['slug'], data['password'])
+    message = game_state.harvest(data['x'], data['y'])
+    game_state.add_recipes()
+    game_state.save(data['slug'])
     return make_response(game_state, message)
 
 
 @app.route('/action/unlock', methods=['GET', 'POST'])
 def unlock():
     data = request.json  # data={slug,x,y,password}
-    game_state = load_state(data['slug'])
-
-    # safety checks
-    if data['password'] != game_state['password']:
-        return "Invalid password", 401
-    plot_price = GAME_CONFIG['plotPrice'] * GAME_CONFIG['plotMultiplier'] ** game_state['unlockCount']
-    if game_state['resources']['cash'] < plot_price:
-        return make_response(game_state, "Not enough cash to unlock plot.")
-
-    # making changes to game_state
-    plot = get_plot(game_state, data['x'], data['y'])
-    plot['locked'] = 0
-    game_state['resources']['cash'] -= plot_price
-    game_state['unlockCount'] += 1
-
-    initialize_state(game_state)  # no money check
-
-    save_state(data['slug'], game_state)
-
-    message = "Unlocked plot for $%s." % plot_price
+    game_state = GameState.load(data['slug'], data['password'])
+    message = game_state.unlock(data['x'], data['y'])
+    game_state.check_cash()
+    game_state.save(data['slug'])
     return make_response(game_state, message)
 
 
 @app.route('/recipe', methods=['POST'])
 def recipe():
-    data = request.json  # data={slug}
-    game_state = load_state(data['slug'])
+    data = request.json  # data={slug,password}
+    game_state = GameState.load(data['slug'], data['password'])
 
     known_recipes = {}
     for recipe_id in RECIPE_CONFIG['recipes']:
-        if recipe_id in game_state['recipes']:
+        if recipe_id in game_state.data['recipes']:
```

Comment:
Instead of accessing `data`, you can hide the implementation inside your class and create an accessor method.

The big reason to do this is to make altering your class behavior later easier.  If you have a method `get_recipes` or `is_known_recipe(recipe_id)` it's clear what it's supposed to do.  Giving general access to the full `data` field outside the class means you have to support whatever way you want to use the class.  In this case, the handler code has to know how game_states are structured, and game_states have to expose this structure so others can manipulate it.

I'd rather have this line read,

`if game_state.is_known_recipe(recipe_id):`

or if that method's not general-purpose enough for you,

`if recipe_id in game_state.get_recipes():`